### PR TITLE
deploy by revision: add --revision to deploy cli and implement.

### PR DESCRIPTION
**Pull request by hmlanigan**
This is part of the implementation of deploy by revision feature work, builds on #13165.

Adds a revision flag to juju deploy and implements its use for charmhub charms.

## QA steps

```console
# hello-juju has revision 6 in the edge channel and revision 8 in the stable channel.
$ juju deploy hello-juju --revision 6 --channel stable
Located charm "hello-juju" in charm-hub, revision 6
Deploying "hello-juju" from charm-hub charm "hello-juju", revision 6 in channel stable

# wait for the unit to settle
$ juju refresh hello-juju

# expected failure, charm deploy requires a channel to refresh too.
$ juju deploy hello-juju --revision 6
ERROR invalid channel for "ch:hello-juju": channel cannot be empty

# deploy a bundle by revision, osm-ha has revision 54 in edge, 37 in candidate and 39 in stable.
# deploying will find the correct bundle, but eventually fail as it isn't a caas model.
$ juju deploy osm-ha --revision 54
Located bundle "osm-ha" in charm-hub, revision 54
...

# expected failure
$ juju deploy kubeflow-lite --revision 53 --channel stable
ERROR revision and channel are mutually exclusive when deploying a bundle. Please choose one.

# ensure deploy a charmstore charm still works
$ juju deploy cs:ubuntu-17
Located charm "ubuntu" in charm-store, revision 17
Deploying "ubuntu" from charm-store charm "ubuntu", revision 17 in channel stable
$ juju refresh ubuntu
Added charm-store charm "ubuntu", revision 18 in channel stable, to the model

# ensure there is no ambiguity related to which charmstore revision to deploy.
$ juju deploy cs:ubuntu-18 --revision 17
ERROR two different revisions to deploy specified: 18 and 17, please choose one.
```

### CODE REVIEW

File: apiserver/facades/client/charms/charmhubrepo.go [hunk 1]
```diff
@@ -91,13 +89,41 @@ func (c *chRepo) ResolveWithPreferredChannel(curl *charm.URL, origin corecharm.O
...
+		effectiveChannel = res.EffectiveChannel
 	}
 
 	// Use the channel that was actually picked by the API. This should
 	// account for the closed tracks in a given channel.
-	channel, err := charm.ParseChannelNormalize(res.EffectiveChannel)
+	channel, err := charm.ParseChannelNormalize(effectiveChannel)
 	if err != nil {
-		return nil, corecharm.Origin{}, nil, errors.Annotatef(err, "invalid channel")
+		return nil, corecharm.Origin{}, nil, errors.Annotatef(err, "heather invalid channel for %q", curl)
```
Comment (achilleasa) [inline apiserver/facades/client/charms/charmhubrepo.go:126]: heather invalid channel? :-)

File: cmd/juju/application/deploy.go [hunk 2]
```diff
@@ -609,6 +612,7 @@ func (c *DeployCommand) SetFlags(f *gnuflag.FlagSet) {
 	f.Var(cmd.NewAppendStringsValue(&c.BundleOverlayFile), "overlay", "Bundles to overlay on the primary bundle, applied in order")
 	f.StringVar(&c.ConstraintsStr, "constraints", "", "Set application constraints")
 	f.StringVar(&c.Series, "series", "", "The series on which to deploy")
+	f.IntVar(&c.Revision, "revision", -1, "The revision to deploy")
```
Comment (achilleasa) [inline cmd/juju/application/deploy.go:615]: Do we additionally need to explain what '-1' means in this context?

**Review by achilleasa — APPROVED** [review]
LGTM and the QA steps worked as expected.

Just a small question about the documentation of the --revision flag.

File: cmd/juju/application/deploy.go (hunk 2 above)
Comment (hmlanigan) [reply cmd/juju/application/deploy.go:615]: I don't think so, -1 as a revision is internal to juju and same as how charm.URL.Revision works.  Although, maybe a check for > -1 is a good idea.

File: apiserver/facades/client/charms/charmhubrepo.go (hunk 1 above)
Comment (hmlanigan) [reply apiserver/facades/client/charms/charmhubrepo.go]: oops, thank you.

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**